### PR TITLE
Simplify and cleanup the `test_mypy.py` file

### PR DESCRIPTION
**Pull request by Viicos**
## Change Summary

Part of https://github.com/pydantic/pydantic/issues/10197

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: tests/mypy/test_mypy.py [hunk 1]
```diff
@@ -35,52 +48,44 @@
...
-            built_cases = [pytest.param(config, module, marks=self.marks) for config, module in built_cases]
-        return built_cases
-
-
-cases = (
+# Type hint taken from the signature of `pytest.param`:
+Marks: TypeAlias = MarkDecorator | Collection[MarkDecorator | Mark]
+
+
+def build_cases(
```
Comment (Viicos) [inline tests/mypy/test_mypy.py:55]: Replaces `MypyCasesBuilder` (no need for a class here) and also enforce providing `list[str]` for the arguments, which simplifies the body to a single list comprehension.

File: tests/mypy/test_mypy.py [hunk 2]
```diff
@@ -157,15 +161,24 @@ def _convert_to_output_path(v: str) -> Path:
...
-            existing = MypyTestTarget(parsed_version, output_path)
+            existing = output_path
             break
 
-    current = MypyTestTarget(MYPY_VERSION_TUPLE, _convert_to_output_path(mypy_version))
-    return MypyTestConfig(existing, current)
+    return MypyTestConfig(existing, _convert_to_output_path(mypy_version))
+
+
+def get_expected_return_code(source_code: str) -> int:
```
Comment (Viicos) [inline tests/mypy/test_mypy.py:170]: Moved from above.

File: pydantic/version.py [hunk 3]
```diff
@@ -63,16 +63,16 @@ def version_info() -> str:
     return '\n'.join('{:>30} {}'.format(k + ':', str(v).replace('\n', ' ')) for k, v in info.items())
 
 
-def parse_mypy_version(version: str) -> tuple[int, ...]:
```
Comment (sydney-runkle) [inline pydantic/version.py:66]: Lol why did we do this before 🤦‍♀️

File: pydantic/version.py [hunk 4]
```diff
@@ -63,16 +63,16 @@ def version_info() -> str:
...
-    like `0.940+dev.04cac4b5d911c4f9529e6ce86a27b44f28846f5d.dirty`.
+    It parses normal version like `1.11.0` and extra info followed by a `+` sign
+    like `1.11.0+dev.d6d9d8cd4f27c52edac1f537e236ec48a01e54cb.dirty`.
 
     Args:
         version: The mypy version string.
 
     Returns:
-        A tuple of ints. e.g. (0, 930).
+        A 3-tuple of ints, e.g. `(1, 11, 0)`.
```
Comment (sydney-runkle) [inline pydantic/version.py:76]: thruple? 🙃

File: pydantic/version.py (hunk 3 above)
Comment (sydney-runkle) [reply pydantic/version.py:66]: Ahhh ok I see I guess it was actually sort of reasonable re new needed ignore comment

File: tests/mypy/test_mypy.py (hunk 1 above)
Comment (sydney-runkle) [reply tests/mypy/test_mypy.py:55]: Great simplification!

File: tests/mypy/test_mypy.py [hunk 5]
```diff
@@ -237,50 +250,29 @@ def test_bad_toml_config() -> None:
...
+        importlib.import_module(module_name)
+    except Exception:
+        pytest.fail(reason=f'Unable to execute module {module_name}')
 
 
 @pytest.mark.parametrize(
-    'v_str,v_tuple',
+    ['v_str', 'v_tuple'],
     [
         ('0', (0,)),
```
Comment (sydney-runkle) [inline tests/mypy/test_mypy.py:270]: Why do we need to test this one? If we don't, we can change the hint below as suggested.

File: tests/mypy/test_mypy.py [hunk 6]
```diff
@@ -237,50 +250,29 @@ def test_bad_toml_config() -> None:
...
-        ('0.930', (0, 930)),
-        ('0.940+dev.04cac4b5d911c4f9529e6ce86a27b44f28846f5d.dirty', (0, 940)),
+        ('1.11.0', (1, 11, 0)),
+        ('1.11.0+dev.d6d9d8cd4f27c52edac1f537e236ec48a01e54cb.dirty', (1, 11, 0)),
     ],
 )
-def test_parse_mypy_version(v_str, v_tuple):
+def test_parse_mypy_version(v_str: str, v_tuple: tuple[int, ...]) -> None:
```
Comment (sydney-runkle) [inline tests/mypy/test_mypy.py:275]:
```suggestion
def test_parse_mypy_version(v_str: str, v_tuple: tuple[int, int, int]) -> None:
```

**Review by sydney-runkle — APPROVED** [review]
LGTM, a few questions 👍

File: tests/mypy/test_mypy.py [hunk 7]
```diff
@@ -35,52 +48,44 @@
...
-            for module in self.modules:
-                built_cases.append((config, module))
-        if self.marks is not None:
-            built_cases = [pytest.param(config, module, marks=self.marks) for config, module in built_cases]
-        return built_cases
-
-
-cases = (
+# Type hint taken from the signature of `pytest.param`:
+Marks: TypeAlias = MarkDecorator | Collection[MarkDecorator | Mark]
```
Comment (sydney-runkle) [inline tests/mypy/test_mypy.py:52]: Need `Union` for 3.8 and 3.9 compat

**Pull request merged**